### ISSUE TITLE
Make the LLVM back end work for dynamic linking on Cray systems

### ISSUE THREAD

**Issue by dmk42** [chapel-lang/chapel#8901]
As a Chapel user, I want to be able to use LLVM with dynamic linking on Cray systems, because some situations require it (such as KNL, which needs the dynamic-only memkind library).

The Chapel interface to the LLVM back end (specifically compiler/util/clangUtil.cpp) uses the script util/config/gather-cray-prgenv-arguments.bash to discover which options to pass to the link step.  However, in the case of dynamic linking, the arguments gathered are not sufficient.  Some libraries are missing, leading to undefined symbols.

This is a unified diff showing the missing library options.

```diff
--- static.opts	2018-03-20 19:26:35.642940029 -0700
+++ dynamic.opts	2018-03-20 19:26:42.476528455 -0700
@@ -280,36 +280,11 @@
 -lre2
 -lpthread
 -Wl,-rpath=/opt/cray/pe/gcc-libs
+-Wl,-Bdynamic
 -L/opt/cray/pe/libsci/18.03.1/GNU/7.1/x86_64/lib
 -L/opt/cray/dmapp/default/lib64
 -L/opt/cray/pe/mpt/7.7.0/gni/mpich-gnu/5.1/lib
 -L/opt/cray/rca/2.2.18-6.0.6.0_19.14__g2aa4f39.ari/lib64
--L/opt/cray/pe/pmi/5.0.13/lib64
--L/opt/cray/xpmem/2.2.14-6.0.6.0_10.1__g34333c9.ari/lib64
--L/opt/cray/ugni/6.0.14-6.0.6.0_18.12__g777707d.ari/lib64
--L/opt/cray/udreg/2.3.2-6.0.6.0_15.18__g5196236.ari/lib64
--L/opt/cray/alps/6.6.0-6.0.6.0_35.25__gd0a1ab9.ari/lib64
--L/opt/cray/wlm_detect/1.3.2-6.0.6.0_3.8__g388ccd5.ari/lib64
--L/opt/cray/alps/6.6.0-6.0.6.0_35.25__gd0a1ab9.ari/lib64
 -Wl,-Ttext-segment=0x20000000,-zmax-page-size=0x20000000
 -Wl,--whole-archive,-lhugetlbfs,--no-whole-archive
--lsci_gnu_71_mpi
--lsci_gnu_71
--lpthread
--lhugetlbfs
--lmpich_gnu_51
--lrt
--lugni
--lpthread
--lpmi
--lpthread
--lalpslli
--lpthread
--lwlm_detect
--lalpsutil
--lpthread
 -lrca
--lxpmem
--lugni
--lpthread
--ludreg
```

**Comment by dmk42** [comment]
@ronawho reminded me to look here for a solution.

https://github.com/chapel-lang/chapel/blob/d0f55d13fe1dcdd56d5f68882d5fa5bf0d3155c2/make/compiler/Makefile.cray-prgenv#L71-L91

**Comment by ronawho** [comment]
Pretty confident adding that logic (which is what https://github.com/chapel-lang/chapel/issues/8901#issuecomment-374836705 was suggesting) will resolve this. 

Static linking on master with --llvm works fine:

```bash
chpl test/release/examples/hello.chpl  --llvm
# compiles fine
```

Dynamic linking on master with --llvm fails:

```bash
export CRAYPE_LINK_TYPE=dynamic
chpl test/release/examples/hello.chpl  --llvm
# lots of undefined references to PMI_ and GNI_ functions
```

Dynamic linking explicitng setting things Makefile.cray-prgenv does today works fine:

```bash
export CRAYPE_LINK_TYPE=dynamic

export PE_PKGCONFIG_PRODUCTS=PE_CHAPEL:$PE_PKGCONFIG_PRODUCTS
export PE_CHAPEL_MODULE_NAME=chapel
export PE_CHAPEL_PKGCONFIG_LIBS=cray-pmi:cray-ugni:$PE_CHAPEL_PKGCONFIG_LIBS

chpl test/release/examples/hello.chpl  --llvm
# compiles fine
```

**Comment by ronawho** [comment]
I was hoping to just split out that Makefile.cray-prgenv logic into a common makefile, and include it in the clang-included makefile as well, but I forgot that we no longer call the makefiles during llvm codegen (as of #7757 I think.)

The quick-and-dirty (but gross because it's duplicating code in the compiler and in makefiles) solution would be something like:

```diff
diff --git a/compiler/util/clangUtil.cpp b/compiler/util/clangUtil.cpp
index fa001285ea..884e58f980 100644
--- a/compiler/util/clangUtil.cpp
+++ b/compiler/util/clangUtil.cpp
@@ -2660,7 +2660,19 @@ void makeBinaryLLVM(void) {
   expandInstallationPaths(clangLDArgs);

   if (compilingWithPrgEnv()) {
-    std::string gather_prgenv(CHPL_HOME);
+    std::string gather_prgenv = "";
+
+    gather_prgenv += "PE_PKGCONFIG_PRODUCTS=PE_CHAPEL:$PE_PKGCONFIG_PRODUCTS ";
+    gather_prgenv += "PE_CHAPEL_MODULE_NAME=chapel ";
+
+    if (strcmp(CHPL_COMM, "ugni") == 0 || strcmp(CHPL_COMM, "ofi") == 0)
+      gather_prgenv += "PE_CHAPEL_PKGCONFIG_LIBS=cray-pmi:cray-ugni:$PE_CHAPEL_PKGCONFIG_LIBS ";
+    else if (strcmp(CHPL_COMM, "gasnet") == 0)
+      if (strcmp(CHPL_COMM_SUBSTRATE, "aries") == 0 || strcmp(CHPL_COMM_SUBSTRATE, "gemini") == 0)
+        gather_prgenv += "PE_CHAPEL_PKGCONFIG_LIBS=cray-pmi:cray-udreg:cray-ugni:$PE_CHAPEL_PKGCONFIG_LIBS ";
+
+    gather_prgenv += CHPL_HOME;
+
     gather_prgenv += "/util/config/gather-cray-prgenv-arguments.bash link";
     readArgsFromCommand(gather_prgenv, clangLDArgs);
   }
```

**Comment by dmk42** [comment]
Closed by #10902
